### PR TITLE
[Frontend] Lock the swiftmodule when rebuilding from the swiftinterface

### PR DESCRIPTION
**Pull request by xymus**
When rebuilding a module interface from the textual interface, lock the destination path of the created swiftmodule instead of the source swiftinterface. The swiftinterface files are likely to be in the SDK and may be on a read-only filesystem.

rdar://60247977

I'm reenabling the test at the same time.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Review by artemcm — APPROVED** [review]
If we have a one-to-many situation where multiple frontends are building the same interface file into multiple binary modules at different paths, we will now spawn more threads and do this in parallel, vs. one-by-one. 

I don't imagine being a problem because we are mostly building into the same location in the Module Cache, but do we know of any other scenarios where this could cause trouble?

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
I think this version is more reliable in this scenario because if a different process was waiting on another one to rebuild the module for a different cache it would have been in vain.

I'd like @nkcsgexi opinion on this, in case I missed something.

**Review by DougGregor — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test macOS

**Comment by xymus** [comment]
@swift-ci Please smoke test Windows

**Review by nkcsgexi — APPROVED** [review]
Thank you for fixing this, @xymus !

**Comment by xymus** [comment]
The APIJSON tests were failing because they share a module cache and use the same name for different modules. They were writing to the the same swiftmodule file which then made some parallel tests read from the swiftmodule generated by another test. This was made more easily detectable with the lock on the swiftmodules instead of the swiftinterfaces. Using local module cache should be enough to avoid this. cc @cachemeifyoucan 

@swift-ci Please smoke test

**Comment by cachemeifyoucan** [comment]
> The APIJSON tests were failing because they share a module cache and use the same name for different modules. They were writing to the the same swiftmodule file which then made some parallel tests read from the swiftmodule generated by another test. This was made more easily detectable with the lock on the swiftmodules instead of the swiftinterfaces. Using local module cache should be enough to avoid this. cc @cachemeifyoucan
> 
> @swift-ci Please smoke test

Thanks for fixing

**Comment by xymus** [comment]
@swift-ci Please smoke test Windows

File: test/Driver/lock_interface.swift [hunk 1]
```diff
@@ -1,23 +1,29 @@
...
+// RUN: %target-swiftc_driver -j20 %t/main.swift %t/file-01.swift %t/file-02.swift %t/file-03.swift -I %t/sdk -Xfrontend -Rmodule-interface-rebuild -module-cache-path %t/module-cache &> %t/result.txt
 // RUN: %FileCheck %s  -check-prefix=CHECK-REBUILD < %t/result.txt
 
-// RUN: %target-swiftc_driver -j20 %t/main.swift %t/file-01.swift %t/file-02.swift %t/file-03.swift -I %t -Xfrontend -Rmodule-interface-rebuild -Xfrontend -disable-interface-lock -module-cache-path %t/module-cache-no-lock &> %t/result.txt
+// RUN: %target-swiftc_driver -j20 %t/main.swift %t/file-01.swift %t/file-02.swift %t/file-03.swift -I %t/sdk -Xfrontend -Rmodule-interface-rebuild -Xfrontend -disable-interface-lock -module-cache-path %t/module-cache-no-lock &> %t/result.txt
 
 // RUN: %FileCheck %s  -check-prefix=CHECK-REBUILD-NO-LOCK < %t/result.txt
 
+// Reset the permissions
+// RUN: chmod a+w %t/sdk
```
Comment (davezarzycki) [inline test/Driver/lock_interface.swift:21]: I think this test is brittle for incremental builds. It really needs a `chmod a+x` at the start of the test to reset %t/sdk to a known good state before running line 5: `echo 'public func foo() {}' > %t/sdk/Foo.swift`

**Pull request merged**